### DISCUSSION TITLE
Caching missing/undefined guild members

### DISCUSSION THREAD

**Discussion by TebexOllie** [discordjs/discord.js#7963] [Q&A] [answered]
Hello! We appear to be hitting rate limits when attempting to fetch members that aren't part of the guild because discord.js understandably doesn't cache missing members. Whilst we can implement this caching ourselves, we wanted to check if there's an existing solution that discord.js provides to essentially cache missing members for a predefined amount of time (e.g. 30 seconds)?

To clarify, the current discord.js functionality is the following:

```js
// Makes API call and returns undefined
let missingMember = await guild.members.fetch(missingId);

// Makes another API call
missingMember = await guild.members.fetch(missingId);
```

However, we're wanting to prevent unnecessary API calls by caching the undefined member for a brief period of time. The end result would look something like the following:

```js
// Makes API call and returns undefined
let missingMember = await guild.members.fetch(missingId);

// Does not make another API call because we already know this member is missing
missingMember = await guild.members.fetch(missingId);
```

As previously mentioned, we can implement this ourselves however I'm sure others must've also encountered issues.

**Comment by Jiralite** [comment]
I'm very confused!

> We appear to be hitting rate limits when attempting to fetch members that aren't part of the guild because discord.js understandably doesn't cache missing members.

How... how can you even fetch members that aren't part of the guild? Discord will reject your API request!

Also, what is a 'missing member'? I'm not sure what this terminology means.

All of your code block samples should never return `undefined`, which is more confusing...

As for the caching part, discord.js doesn't have a default limit on caching guild members. If you are losing guild members, that means you are sweeping them out of the cache (and that means you know you are doing it).

**Reply by TebexOllie** [reply]
Thanks for the reply! We want to cache the fact that the member doesn't exist in the guild, so that we aren't making a load of API calls to Discord trying to fetch a member we already know doesn't exist. Being able to cache that `undefined` result for a short period of time is what we're wanting to achieve, so if discord.js has a way of doing so that would be great, although I appreciate this is quite an unusual aspiration.

**Reply by Jiralite** [reply]
How do you know of a user that may or may not be in the guild? Do you have a database of user ids you are wanting to check if they are in the guild?

**Reply by TebexOllie** [reply]
Yes that's correct, we have a DB containing user IDs.

**Comment by TebexOllie** [comment, marked as answer]
We've since implemented on our side to prevent unnecessary API calls.
